### ISSUE TITLE
zerotierone: can no longer reach cloud based VM after 19.03→19.09 update

### ISSUE THREAD

**Issue by obadz** [NixOS/nixpkgs#71955]
**Describe the bug**
I use zerotier as a "sort of" VPN layer to communicate with my cloud VMs. After updating to 19.09, I'm no longer able to ping the VM via zerotier.

**To Reproduce**
1. Install a VM on Vultr (not sure it would be the same somewhere else)
2. Configure zerotierone on it
3. Set up a zerotier network including another box (not on Vultr)
4. Hosts can no longer ping each other

Reverting to 19.03 fixes the issue.
Reverting _just_ zerotier to the same version as 19.03 *DOES NOT* fix the issue. (The package was updated. The module hasn't changed significantly.).

Maybe some firewall rule or other network setting?

I attempted a git bisect over nixpkgs but the problem seemed to no longer be reproducible when I tried in rapid succession which led me to distrust any information I got out of that bisect and I eventually gave up.

**Expected behavior**
Would expect the pings to get through.

Curious to hear if anyone else is having the issue?

**Additional context**
(My other boxes which aren't cloud based and use NixOS unstable can ping each other just fine).

Ironically the cloud VM is the only box with a public IP in this story so it should have the easiest time opening tunnels

`zerotier-cli listpeers` seems to show the boxes I'm trying to reach, so I'd guess something is going wrong somewhere near the tun/tap level rather than with underlying transport.

**Metadata**
0d79decdb3b was fine
936a53ee699 to at least 8bf142e001b are broken

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by danielfullmer** [comment]
I believe the source of this issue is: fe3da83b7e26b6f7cdde7a457a794c215d16e969 (this commit under 19.09 branch: b4c2e6b6c50008a4e6c35f44a36285fca07bca8d)
This commit loads the `tun` module at boot if `boot.enableContainers` is enabled (which it is by default).

I found this by bisecting nixpkgs with my own nixos test for zerotier connectivity: https://github.com/danielfullmer/nixos-config/blob/master/tests/zerotier/default.nix
It fails while trying to ping the remote computer as well.
I just did a workaround is to set `boot.enableContainers = false` for the time being:
https://github.com/danielfullmer/nixos-config/commit/ba91641c7ae54525182b65460c82fb60d73c8d29

I was considering making an issue for this--but since I could not reproduce this issue in any of my physical machines I wasn't sure if it was a genuine issue, or just an issue with my custom test.

It's interesting that it fails on vultr as well as under nixos tests/qemu, but not on physical machines. I can't think of a reason why loading "tun" early would cause this, and I don't have any great idea about how to continue debugging the underlying cause.

**Comment by Dbone29** [comment]
I had the same Issue on a server from https://netcup.de.
I could reproduce the issue multiple times.
"boot.enableContainers = false" fixed it.

**Comment by obadz** [comment]
I'm afraid this didn't fix the problem for me :-(

**Comment by danielfullmer** [comment]
@obadz @Dbone29  I may have fixed the underlying issue here: https://github.com/danielfullmer/nixos-config/blob/master/tests/zerotier/default.nix

It turns out that the MAC address was being changed by systemd after it was set by zerotier. You can compare the MAC address shown via "ip link" with what it should be on [my.zerotier.com](https://my.zerotier.com/)

**Comment by obadz** [comment]
Closed by #77405. Thanks @danielfullmer!
